tachyon/crypto/commitments/fri/two_adic_fri_pcs_unittest.cc

```diff
@@ -0,0 +1,196 @@
+    Poseidon2Sponge<Poseidon2ExternalMatrix<Poseidon2Plonky3ExternalMatrix<F>>>;
+using PackedPoseidon2 = Poseidon2Sponge<
+    Poseidon2ExternalMatrix<Poseidon2Plonky3ExternalMatrix<PackedF>>>;
+using MyHasher = PaddingFreeSponge<Poseidon2, kRate, kChunk>;
+using MyPackedHasher = PaddingFreeSponge<PackedPoseidon2, kRate, kChunk>;
+using MyCompressor = TruncatedPermutation<Poseidon2, kChunk, kN>;
+using MyPackedCompressor = TruncatedPermutation<PackedPoseidon2, kChunk, kN>;
+using MMCS = FieldMerkleTreeMMCS<F, MyHasher, MyPackedHasher, MyCompressor,
+                                 MyPackedCompressor, 8>;
+using ExtMMCS = FieldMerkleTreeMMCS<ExtF, MyHasher, MyPackedHasher,
+                                    MyCompressor, MyPackedCompressor, 8>;
+using ChallengeMMCS = ExtensionFieldMerkleTreeMMCS<ExtF, ExtMMCS>;
+using Challenger = zk::air::plonky3::DuplexChallenger<Poseidon2, 16, 8>;
+using MyPcs = TwoAdicFriPCS<ExtF, MMCS, ChallengeMMCS, Challenger>;
+
+class TwoAdicFriPCSTest : public testing::Test {
+ public:
+  TwoAdicFriPCSTest() = default;
+
+  static void SetUpTestSuite() {
+    ExtF::Init();
+    ExtendedPackedF::Init();
+  }
+
+  void SetUp() override {
+    Poseidon2Config<F> config = Poseidon2Config<F>::CreateCustom(
+        15, 7, 8, 13, math::GetPoseidon2BabyBearInternalShiftVector<15>());
+    Poseidon2 sponge(config);
+    MyHasher hasher(sponge);
+    MyCompressor compressor(sponge);
+
+    Poseidon2Config<PackedF> packed_config =
+        Poseidon2Config<PackedF>::CreateCustom(
+            15, 7, 8, 13, math::GetPoseidon2BabyBearInternalShiftVector<15>());
+    PackedPoseidon2 packed_sponge(packed_config);
+    MyPackedHasher packed_hasher(packed_sponge);
+    MyPackedCompressor packed_compressor(std::move(packed_sponge));
+    MMCS mmcs(hasher, packed_hasher, compressor, packed_compressor);
+
+    ChallengeMMCS challenge_mmcs(
+        ExtMMCS(std::move(hasher), std::move(packed_hasher),
+                std::move(compressor), std::move(packed_compressor)));
+
+    // TODO(ashjeong): Include separate test for |log_blowup| = 2
+    TwoAdicFriConfig<ChallengeMMCS> fri_config{1, 10, 8, challenge_mmcs};
+
+    pcs_ = MyPcs(std::move(mmcs), std::move(fri_config));
+    challenger_ = Challenger(std::move(sponge));
+  }
+
+  void TestProtocol(std::vector<std::vector<size_t>> log_degrees_by_round) {
+    using Commitment = typename MMCS::Commitment;
+    using ProverData = typename MMCS::ProverData;
+    using OpenedValues =
+        std::vector<std::vector<std::vector<std::vector<ExtF>>>>;
+    using Proof =
+        TwoAdicFriProof<ChallengeMMCS, std::vector<BatchOpening<MMCS>>, F>;
+    using Claims = std::vector<std::tuple<ExtF, std::vector<ExtF>>>;
+
+    size_t num_rounds = log_degrees_by_round.size();
+    std::vector<std::vector<Domain>> domains_by_round(num_rounds);
+    std::vector<Commitment> commits_by_round(num_rounds);
+    std::vector<ProverData> data_by_round(num_rounds);
+    Challenger p_challenger = challenger_;
```

Comment:
Why do you need this?

Reply:
`Challenger v_challenger = challenger_;` occurs later on as while both the prover and verifier challenger start the same, they differ in later stages.